### ISSUE TITLE
Use dynamic client instead of rest client

### ISSUE THREAD

**Issue by wlynch** [tektoncd/triggers#233]
# Proposal

Currently, we use the Kubernetes REST client for all resource creation. Instead, I think we should switch to using the [dynamic client](https://godoc.org/k8s.io/client-go/dynamic).

Our current REST client implementation assumes that the underlying resource is meant for Kubernetes. By abstracting this to the dynamic client, we get a few benefits:

1. We can replace a majority of our REST client code with the dynamic client, which already has good support for CRDs. See https://soggy.space/namespaced-crds-dynamic-client/ for a simple example.
2. We can take advantage of the [fake](https://godoc.org/k8s.io/client-go/dynamic/fake) client for use in tests.
2. This makes supporting other operations like #231 for CRDs easy, since the dynamic client works pretty seamlessly for any registered resource.
3. Providers can bring their own types by following the dynamic interface. I was able to get a nifty demo working where Tekton Triggers were creating GCB Builds, but hypothetically this can be done for any k8s or external resource, so long it satisfies the interface.

Demo/PoC: https://github.com/tektoncd/triggers/compare/master...wlynch:dynamic-client#diff-57fed74cee793226b40869dac1113d50R194-R199 (WIP)

### tl;dr 

Turn this:
```go
uri := createRequestURI(apiVersion, apiResource.Name, namespace, apiResource.Namespaced)
result := r.RESTClient.Post().
	RequestURI(uri).
	Body([]byte(rt)).
	SetHeader("Content-Type", "application/json").
	Do()
```

Into this:
```go
gvr := schema.GroupVersionResource{
	Group:    apiResource.Group,
	Version:  apiResource.Version,
	Resource: apiResource.Name,
}
_, err = r.DynamicClient.Resource(gvr).Namespace(namespace).Create(data, metav1.CreateOptions{})
```

**Comment by sbose78** [comment]
+1 on using the dynamic client. We use it everywhere in https://github.com/redhat-developer/service-binding-operator
